File: the-order-arguments-in-left-folds/javascript/reduce.js
const rev1 = () => {
  const collection = [1, 2, 3, 4, 5];
  const initialAccumulator = [];
  const combiningFunction = (currentAccumulator, currentItem) => [
    currentItem,
    ...currentAccumulator,
  ];

  return collection.reduce(combiningFunction, initialAccumulator);
};

const reduceWithNamedArgs = ({ over, from, withFn }) => {
  return over.reduce(withFn, from);
};

const rev2 = () => {
  const collection = [1, 2, 3, 4, 5];
  const initialAccumulator = [];
  const combiningFunction = (currentAccumulator, currentItem) => [
    currentItem,
    ...currentAccumulator,
  ];

  return reduceWithNamedArgs({
    over: collection,
    from: initialAccumulator,
    withFn: combiningFunction,
  });
};

console.log(rev1());
console.log(rev2());

File: the-order-arguments-in-left-folds/elixir/lib/reduce.ex
defmodule Reduce do
  def rev do
    collection = [1, 2, 3, 4, 5]
    initial_accumulator = []

    combining_function = fn current_item, current_accumulator ->
      [current_item | current_accumulator]
    end

    Enum.reduce(
      collection,
      initial_accumulator,
      combining_function
    )
  end

  def run do
    IO.inspect(rev())
  end
end
